### ISSUE TITLE
[sklearn#34078] BUG: Decision trees: `tree.tree_.impurity` reports incorrect impurity for monotonic trees

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#354]
**Mirrored from [scikit-learn#34078](https://github.com/scikit-learn/scikit-learn/issues/34078)**

Original title: BUG: Decision trees: `tree.tree_.impurity` reports incorrect impurity for monotonic trees
Original labels: Needs work
Created: 2026-05-21T16:12:55Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the bug

For monotonic decision trees, `tree.tree_.impurity` can correspond to the impurity value computed before clipping. This makes `tree.tree_.impurity` inconsistent with `tree.tree_.value`.

In the reproducer below, one leaf contains only targets equal to 0, but its prediction is clipped to 0.25 to satisfy the monotonic constraint. `tree.tree_.impurity`/`plot_tree` still reports `squared_error = 0.0` for that leaf, even though the squared error around the actual prediction is 0.0625.

### Steps/Code to Reproduce

```python
import numpy as np
from sklearn.tree import DecisionTreeRegressor, plot_tree

X = np.array([
    [0, 1, 2, 3, 4, 5, 6, 7],
    [0, 1, 2, 3, 0, 2, 4, 6],
]).T
y = [0, 0, 0, 0, 1, 1, 0, 0]

tree = DecisionTreeRegressor(monotonic_cst=[1, 0]).fit(X, y)
plot_tree(tree)
```

<img width="515" height="389" alt="Image" src="https://github.com/user-attachments/assets/98fd9641-fe46-4dbf-b7cf-8cd8f29d549a" />

### Expected Results

The impurity shown by `plot_tree` (and in `tree.tree_.impurity`) should be consistent with the actual node value.

### Actual Results

`plot_tree` displays `value = 0.25` and `squared_error = 0.0` for the same leaf (in which all samples have value 0).

### Versions

Observed on the current development checkout of scikit-learn.

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [65](https://github.com/godofecht/flow-scikit/issues/65) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
